app/presenters/facet/range_presenter.rb

```diff
@@ -200,25 +200,48 @@ def displayable_end_value(items)
     def padded_items(items)
       @padded_items ||= begin
         items = items.sort_by(&:value)
-        padded = []
+        padded = pad_items(items)
+        padded = pre_pad_items(padded, items)
+        padded = post_pad_items(padded, items)
+        padded
+      end
+    end
 
-        items.each_with_index do |item, i|
-          padded << item
-          next if item == items.last
+    def pad_items(items)
+      padded = []
 
-          next_item = items[i + 1]
+      items.each_with_index do |item, i|
+        padded << item
+        next if item == items.last
 
-          (item.value..next_item.value).each do |value|
-            next if [item.value, next_item.value].include?(value)
-            padded << value
-          end
+        next_item = items[i + 1]
+
+        (item.value..next_item.value).each do |value|
+          next if [item.value, next_item.value].include?(value)
+          padded << value
         end
+      end
 
-        padded
+      padded
+    end
+
+    def pre_pad_items(padded, items)
+      padded_begin_value = padding_item?(padded.first) ? padded.first : padded.first.value
+      (displayable_begin_value(items)..padded_begin_value).each do |pre_pad|
```

Comment:
This throws an error when there's nothing returned by the facets: 
'collections/migration?q=&range[YEAR][begin]=1001&range[YEAR][end]=1056'

also it seems more logical to just prepend the array of all the values to add in one go rather than unshifting them onto the array one by one. Something like: 
```
(displayable_begin_value(items)..(padded_begin_value-1)).to_a + padded
```

Although that said I tried to benchmark this and unshifting seems slightly faster bar one outlier:
```
irb(main):024:0> Benchmark.measure{ arr2 = (1..999).to_a; x = arr2 + arr1 }
=> #<Benchmark::Tms:0x0055f21da80ab0 @label="", @real=0.0004307909985072911, @cstime=0.0, @cutime=0.0, @stime=0.0, @utime=0.0, @total=0.0>
irb(main):025:0> Benchmark.measure{ arr2 = (1..999).to_a; x = arr2 + arr1 }
=> #<Benchmark::Tms:0x0055f21daa3bf0 @label="", @real=0.0003467010101303458, @cstime=0.0, @cutime=0.0, @stime=0.0, @utime=0.0, @total=0.0>
irb(main):026:0> Benchmark.measure{ arr2 = (1..999).to_a; x = arr2 + arr1 }
=> #<Benchmark::Tms:0x0055f21dabcc68 @label="", @real=0.00044454302405938506, @cstime=0.0, @cutime=0.0, @stime=0.0, @utime=0.0, @total=0.0>
irb(main):027:0> Benchmark.measure{ arr2 = (1..999).to_a; x = arr2 + arr1 }
=> #<Benchmark::Tms:0x0055f21b82d0a8 @label="", @real=0.00034313800279051065, @cstime=0.0, @cutime=0.0, @stime=0.0, @utime=0.0, @total=0.0>
irb(main):028:0> Benchmark.measure{ arr2 = (1..999).to_a; x = arr2 + arr1 }
=> #<Benchmark::Tms:0x0055f21daf5f18 @label="", @real=0.00022153998725116253, @cstime=0.0, @cutime=0.0, @stime=0.0, @utime=0.0, @total=0.0>
irb(main):029:0> Benchmark.measure{ arr2 = (1..999).to_a; x = arr2 + arr1 }
=> #<Benchmark::Tms:0x0055f21b810e58 @label="", @real=0.00033797998912632465, @cstime=0.0, @cutime=0.0, @stime=0.0, @utime=0.0, @total=0.0>
irb(main):030:0> Benchmark.measure{ arr2 = (1..999).to_a; x = arr2 + arr1 }
=> #<Benchmark::Tms:0x0055f21db2e430 @label="", @real=0.00037676398642361164, @cstime=0.0, @cutime=0.0, @stime=0.0, @utime=0.0, @total=0.0>
irb(main):031:0> Benchmark.measure{ arr2 = (1..999).to_a; x = arr2 + arr1 }
=> #<Benchmark::Tms:0x0055f2199d23b0 @label="", @real=0.00041920796502381563, @cstime=0.0, @cutime=0.0, @stime=0.0, @utime=0.0, @total=0.0>
irb(main):032:0> Benchmark.measure{ arr2 = (1..999).to_a; x = arr2 + arr1 }
=> #<Benchmark::Tms:0x0055f21db69378 @label="", @real=0.00023094797506928444, @cstime=0.0, @cutime=0.0, @stime=0.0, @utime=0.0, @total=0.0>
irb(main):033:0> Benchmark.measure{ x = (1..999).each { |val| arr1.unshift(val)} }
=> #<Benchmark::Tms:0x0055f21db8ab90 @label="", @real=0.00026911799795925617, @cstime=0.0, @cutime=0.0, @stime=0.0, @utime=0.0, @total=0.0>
irb(main):034:0> Benchmark.measure{ x = (1..999).each { |val| arr1.unshift(val)} }
=> #<Benchmark::Tms:0x0055f21998da30 @label="", @real=0.00019771402003243566, @cstime=0.0, @cutime=0.0, @stime=0.0, @utime=0.0, @total=0.0>
irb(main):035:0> Benchmark.measure{ x = (1..999).each { |val| arr1.unshift(val)} }
=> #<Benchmark::Tms:0x0055f21db9dcb8 @label="", @real=0.0005558169796131551, @cstime=0.0, @cutime=0.0, @stime=0.0, @utime=0.0, @total=0.0>
irb(main):036:0> Benchmark.measure{ x = (1..999).each { |val| arr1.unshift(val)} }
=> #<Benchmark::Tms:0x0055f2199784a0 @label="", @real=0.0003853610251098871, @cstime=0.0, @cutime=0.0, @stime=0.0, @utime=0.0, @total=0.0>
irb(main):037:0> Benchmark.measure{ x = (1..999).each { |val| arr1.unshift(val)} }
=> #<Benchmark::Tms:0x0055f21d851460 @label="", @real=0.0003062079777009785, @cstime=0.0, @cutime=0.0, @stime=0.0, @utime=0.0, @total=0.0>
irb(main):038:0> Benchmark.measure{ x = (1..999).each { |val| arr1.unshift(val)} }
=> #<Benchmark::Tms:0x0055f21b7ec8c8 @label="", @real=8.654501289129257e-05, @cstime=0.0, @cutime=0.0, @stime=0.0, @utime=0.0, @total=0.0>
irb(main):039:0> Benchmark.measure{ x = (1..999).each { |val| arr1.unshift(val)} }
=> #<Benchmark::Tms:0x0055f21dbcc7e8 @label="", @real=0.0002909339964389801, @cstime=0.0, @cutime=0.0, @stime=0.0, @utime=0.0, @total=0.0>
irb(main):040:0> Benchmark.measure{ x = (1..999).each { |val| arr1.unshift(val)} }
=> #<Benchmark::Tms:0x0055f21dbd6ab8 @label="", @real=0.00025587802520021796, @cstime=0.0, @cutime=0.0, @stime=0.0, @utime=0.0, @total=0.0>
irb(main):041:0> Benchmark.measure{ x = (1..999).each { |val| arr1.unshift(val)} }
=> #<Benchmark::Tms:0x0055f21b7be1a8 @label="", @real=0.00014014600310474634, @cstime=0.0, @cutime=0.0, @stime=0.0, @utime=0.0, @total=0.0>

```
maybe if the arrays were bigger or a larger sample size of test runs would start making a difference.

Reply:
> This throws an error when there's nothing returned by the facets:
'collections/migration?q=&range[YEAR][begin]=1001&range[YEAR][end]=1056'

True, and handled.

> also it seems more logical to just prepend the array of all the values to add in one go rather than unshifting them onto the array one by one. Something like:
> `(displayable_begin_value(items)..(padded_begin_value-1)).to_a + padded`

True, and done, but not quite that way as that would limit the range facet presenter to working with integers, hence the slightly ungainly use of pre/post ranges.